### ISSUE TITLE
How to accelerate the policy training

### ISSUE THREAD

**Issue by gyou2021** [Garfield-kh/PoseTriplet#23]
Hi, I changed the loss function of the training for the 3D pose estimation. So I should train the policy since the iteration helix_1 and I cannot use the policy trained by you, right?
However, I found it is time-consuming to train the policy (
python pose_imitation/pose_mimic.py --cfg subject_h36m_helix_1 --num-threads 52 --mocap-folder ./checkpoint/exp_h36m_gt2d_v5/helix_1). About 11 hours were needed on my server to get a model such as models/iter_0200.p. Would you please tell me how to accelerate the policy training process? Thank you!

**Comment by Garfield-kh** [comment]
Hi, sorry I didn't understand the first question.

For the acceleration, the imitator is built based on Mujoco, which is a CPU based simulator. So the speed of training is limited by the "--num-threads 52" and the actual CPU CORE number. I run it on a 27x2 core CPU so I can get 100 iter/hour for the training.
You may try if this **export OMP_NUM_THREADS=1** helps as suggested from [RFC](https://github.com/Khrylx/RFC).
Or try on a machine with strong CPU.
Another solution is to change the simulator from Mujoco to [isaac-gym](https://developer.nvidia.com/isaac-gym), which is GPU based simulator, with hundreds of speed up, but this requires a lot of code work.

**Comment by gyou2021** [comment]
Hi,  it works export OMP_NUM_THREADS=1. The training time for 100 iterations was reduced to about 50 minutes. (CPU 2 sockets, 28 cores per socket, 2500 MHZ). Thank you!
The first question is: you have provided pre-trained imitator models such as ./checkpoint/exp_h36m_gt2d_v5/helix_1/results/posemimic/subject_h36m_helix_1/results/iter_4000_train_naivefs.p, however, I have changed the loss function of training the 3D pose estimation in training estimator with GAN projection, (the classic loss function is mpjpe, in posegan_train.py MSE was used to get adv_cam_loss). So I will get an estimator model different from yours and then get a different traj_dict.pkl. with temp bvh file, then generate a different expert file. Based on the expert file, the result of training the policy will be different from the pre-trained model you provided such as iter_4000_train_naivefs.p, right?  I cannot use the pre-trained model of policy training (posemimic.py) you provided to train estimator with GAN projection (posegan_train.py) because I changed the loss function in 3D pose training.

**Comment by Garfield-kh** [comment]
Hi, nice to see `OMP_NUM_THREADS=1` works!
`./checkpoint/exp_h36m_gt2d_v5/helix_1/results/posemimic/subject_h36m_helix_1/results/iter_4000_train_naivefs.p` is not pre-trained imitator models, but physical corrected motion result which is estimated from estimator of helix_0 [line 34](https://github.com/Garfield-kh/PoseTriplet/blob/8d1cfa0037923a26e510d03312916345b8d2a401/imitator/script-summary-gt2d-v5.sh#L34) and physical corrected by imitator fitting/train [line 56](https://github.com/Garfield-kh/PoseTriplet/blob/8d1cfa0037923a26e510d03312916345b8d2a401/imitator/script-summary-gt2d-v5.sh#L56).

It seems you have changed the esimator module and loss design, so the whole data fitting/estimation/correction path will change except the first step, random data generation untill [line 17](https://github.com/Garfield-kh/PoseTriplet/blob/8d1cfa0037923a26e510d03312916345b8d2a401/imitator/script-summary-gt2d-v5.sh#L17).

By my experience, the `take_599 Hip flip` issue is the one requires a proper initialization. You can continue your training process if that issue is not occurs.
One more thing is that you may keep an eyes on the GPU-based [isaac-gym](https://developer.nvidia.com/isaac-gym) to see if any motion-imitation work avaliable. That one can reduce RL training time from 2 days to 30 mins for [DeepMimic](https://github.com/xbpeng/DeepMimic)/[AMP](https://xbpeng.github.io/projects/AMP/index.html). Highly recommended.

**Comment by gyou2021** [comment]
Hi, Thank you for your detailed explanation!  
Only _train_batch_posenet() for training and evaluate_posenet() for evaluation in posegan_basementclass.py are needed to be modified when changing the loss function, right?

**Comment by gyou2021** [comment]
Hi, how do you select the value such as 4000 for the parameter --iter when >>inference the RL result using trained model  python pose_imitation/pose_mimic_eval.py --cfg subject_h36m_helix_1 --data train --num-threads 52 --iter 4000 --mocap-folder ./checkpoint/exp_h36m_gt2d_v5/helix_1
Thank you!

**Comment by Garfield-kh** [comment]
There is a [adv_loss](https://github.com/Garfield-kh/PoseTriplet/blob/eb93132f99161bd776dafbcb713e9fb43e501c36/estimator/posegan_train.py#L192) for Augmented projection data augmentation.
The number 4000 is selected from [2000, 4000, 6000], 
In the early try, I think the model weight of eatimator/imitator from last round can be used as pretrained for this round (2000) to save time, but it didn't work well, the reason might be some failure bias from last round will be brought to this round by the model weight. 
For [4000, 6000], this is selected by watching the reward/loss cruve from the imitator training, it shows not much difference between them.

**Comment by gyou2021** [comment]
Thank you for your explanation!
